File: Rapptr_iOS_Test/iOS Programmer Test/iOSTest/ChatClient.m
//
//  ChatClient.m
//  iOSTest
//
//  Created by Sajan Shrestha on 2/11/21.
//  Copyright © 2021 D&ATechnologies. All rights reserved.
//

#import "ChatClient.h"

@interface ChatClient ()
@property (nonatomic, strong) NSURLSession *session;
@end

@implementation ChatClient

/**
 * =========================================================================================
 * INSTRUCTIONS
 * =========================================================================================
 * 1) Make a request to fetch chat data used in this app.
 *
 * 2) Using the following endpoint, make a request to fetch data
 *    URL: http://dev.rapptrlabs.com/Tests/scripts/chat_log.php
 **/

- (void)fetchChatData:(void (^)(NSArray<Message *> *))completion withError:(void (^)(NSString *error))errorBlock
{
    NSString *urlString = @"http://dev.rapptrlabs.com/Tests/scripts/chat_log.php";
    NSURL *url = [NSURL URLWithString:urlString];
    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        
        NSError *err;
        NSDictionary *messageResponse = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];
        
        if (err) {
            NSLog(@"Failed to serialize into JSON: %@", err);
            return;
        }
        
        NSArray *fetchedData  = messageResponse[@"data"];
        NSMutableArray<Message *> *messages = NSMutableArray.new;
        
        for (NSDictionary *fetchedMessage in fetchedData) {
            Message *message = [[Message alloc] initWithDictionary:fetchedMessage];
            [messages addObject:message];
        }
        
        completion(messages);
        
    }] resume];
}

@end

File: Rapptr_iOS_Test/iOS Programmer Test/iOSTest/ChatClient.swift
//
//  ChatClient.swift
//  iOSTest
//
//  Created by Sajan Shrestha on 2/10/21.
//  Copyright © 2021 D&ATechnologies. All rights reserved.
//

import Foundation

struct ChatClient {
    
    static func getMessages(completion: @escaping (([Message]) -> Void)) {
        
        guard let url = URL(string: Endpoint.chatMessagesUrl) else { return }
        
        URLSession.shared.dataTask(with: url) { (data, response, error) in
            guard let data = data else { return }
            do {
                let messageResponse = try JSONDecoder().decode(MessageResponse.self, from: data)
                completion(messageResponse.data)
            }
            catch {
                print("error fetching data")
            }
        }.resume()
    }
}

struct Endpoint {
    static let chatMessagesUrl = "http://dev.rapptrlabs.com/Tests/scripts/chat_log.php"
    static let loginUrl = "http://dev.rapptrlabs.com/Tests/scripts/login.php"
}
